language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -51,8 +51,24 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         let locals_signature_view = function_definition_view.locals_signature();
         let function_signature_view = function_definition_view.signature();
         let mut locals = BTreeMap::new();
+        if function_signature_view.arg_count() > locals_signature_view.len() {
+            return vec![VMStatus::new(StatusCode::RANGE_OUT_OF_BOUNDS)
+                .with_message("Fewer locals than parameters".to_string())];
+        }
+        let mut errors = vec![];
         for (arg_idx, arg_type_view) in function_signature_view.arg_tokens().enumerate() {
-            if arg_type_view.is_reference() {
+            if arg_type_view.as_inner()
+                != locals_signature_view
+                    .token_at(arg_idx as LocalIndex)
```

Comment:
small nit, bind the lhs and rhs to locals to make it a bit more readable (it will take 3 lines instead of 5).